187057225 Setup Data Interactive Handlers

tealefristoe opened the pull request:
PT Story: https://www.pivotaltracker.com/story/show/187057225

This PR establishes infrastructure of using handlers to respond to requests from CODAP v3 plugins.
- Handlers are registered using `registerDIHandler()` in `data-interactive-handler.ts`.
- Handler files should be imported in `register-handlers.ts`.
- Requests from plugins are interpreted using functions defined in `resource-parser.ts`. These functions are closely based on functions from v2 and currently have many commented lines that will be addressed in future PRs.
- `useDataInteractiveController` now uses `getDIHandler()` to find an appropriate registered handler to respond to plugin requests, instead of having these handlers hard coded in the hook.
- Interactive frame handling functions have been moved into a handler defined in `interactive-frame-handler.ts`.
- The get attribute handling function has been created in a handler defined in `attribute-handler.ts`.
  - This required a number of changes to be made to the `Attribute`, `Collection`, and `DataSet` models.
  - It's unclear what the proper responses are for a number of attribute values in v3. TODO comments have been added to indicate values that should be addressed sometime in the future.
- `data-interactive-types.ts` have been further fleshed out.

kswenson reviewed (changes requested):
👍 Generally looks good -- Before recommending some of the changes I wanted to verify that they actually made sense locally, so I've added a commit that implements some of what I'm thinking. Let me know what you think. In other places I made suggestions without trying them out locally. Take a look and we can get together to discuss if further consideration is warranted.

tealefristoe commented on v3/src/models/data/attribute.ts:
```diff
@@ -313,6 +313,31 @@ export const Attribute = types.model("Attribute", {
   }
 }))
+.views(self => ({
+  get toArchive() {
```
Creating a new model was causing a MST error about duplicate ids. I just created the return object directly in the attribute handler, which works and keeps this CODAP specific code out of the attribute model. This might be trickier to do when it comes to updating, though, since we'll want to update everything in a single action that's `withoutUndo`.

kswenson replied:
I wouldn't have thought that creating a local `SnapshotProcessor` would trigger such a warning, but I didn't try it locally. 🤔

tealefristoe commented on v3/src/models/data/data-set.ts:
```diff
@@ -279,6 +279,18 @@ export const DataSet = types.model("DataSet", {
     return collectionId === self.ungrouped.id ? self.ungrouped : getGroupedCollection(collectionId)
   }
 
+  function getGroupedCollectionByName(name: string): ICollectionModel | undefined {
+    return self.collections.find(collection => collection.name === name)
+  }
+
+  function getCollectionByName(name: string): ICollectionPropsModel | undefined {
+    if (!isAlive(self)) {
+      console.warn("DataSet.getCollectionByName called on a defunct DataSet")
+      return
```
I didn't give this too much thought and just went with the path of least resistance--I mimicked `getGroupedCollection` and `getCollection` which are defined right above these, then used `getGroupedCollection` and `getGroupedCollectionByName` because they return `ICollectionModel` instead of `ICollectionPropsModel`.

tealefristoe commented:
@kswenson Thanks for all the feedback. I think I've addressed most of your suggestions. These are the things I didn't address:
- As I mentioned above, I changed `toArchive` to a function directly in the handler, since using a MST model was causing me problems.
- I left `DIValues` as is, though I agree that it will become unwieldy. I also think there might be some cyclical imports we'll need to deal with in my next PR.
- I didn't change the calls to `getGroupedCollection` and `getGroupedCollectionByName`. As I mentioned in the comment above, I didn't give this much thought, so maybe it would make sense to make new functions that include the ungrouped collections and return complete `ICollectionModels`.

kswenson reviewed (approved):
👍 Looks good -- I added a commit to tweak the collection-handling code but there's likely to be more work to do in this area.
